pulsar-common/src/main/java/com/yahoo/pulsar/common/util/collections/GrowablePriorityLongPairQueue.java

```diff
@@ -0,0 +1,404 @@
+import java.util.concurrent.locks.ReentrantLock;
+
+import io.netty.util.internal.MathUtil;
+
+/**
+ * An unbounded priority queue based on a min heap where values are composed of pairs of longs.
+ *
+ * When the capacity is reached, data will be moved to a bigger array.
+ * 
+ * <b>It also act as a set and doesn't store duplicate values if {@link #allowedDuplicate} flag is passed false</b>
+ * 
+ ** <p>
+ * (long,long)
+ * <p>
+ *
+ */
+public class GrowablePriorityLongPairQueue {
+
+    private final ReentrantLock lock = new ReentrantLock();
+
+    private long[] data;
+    private int capacity;
+    private static final AtomicIntegerFieldUpdater<GrowablePriorityLongPairQueue> SIZE_UPDATER = AtomicIntegerFieldUpdater
+            .newUpdater(GrowablePriorityLongPairQueue.class, "size");
+    private volatile int size = 0;
+    private static final long EmptyItem = -1L;
+
+    public GrowablePriorityLongPairQueue() {
+        this(64);
+    }
+
+    public GrowablePriorityLongPairQueue(int initialCapacity) {
+        checkArgument(initialCapacity > 0);
+        this.capacity = MathUtil.findNextPositivePowerOfTwo(initialCapacity);
+        data = new long[2 * capacity];
+        fillEmptyValue(data, 0, data.length);
+    }
+
+    public interface LongPairPredicate {
+        boolean test(long v1, long v2);
+    }
+
+    public static interface LongPairConsumer {
+        void accept(long v1, long v2);
+    }
+
+    public boolean add(long item1, long item2) {
+        lock.lock();
+
+        try {
+
+            if (SIZE_UPDATER.get(this) >= this.capacity) {
+                expandArray();
+            }
+
+            int lastIndex = SIZE_UPDATER.get(this) << 1;
+            data[lastIndex] = item1;
+            data[lastIndex + 1] = item2;
+
+            int loc = lastIndex;
+
+            // Swap with parent until parent not larger
+            while (loc > 0 && compare(loc, parent(loc)) < 0) {
+                swap(loc, parent(loc));
+                loc = parent(loc);
+            }
+
+            SIZE_UPDATER.incrementAndGet(this);
+        } finally {
+            lock.unlock();
+        }
+
+        return true;
+    }
+
+    public void forEach(LongPairConsumer processor) {
+        lock.lock();
+        try {
+            int size = SIZE_UPDATER.get(this);
+
+            int index = 0;
+            for (int i = 0; i < size; i++) {
+                processor.accept(data[index], data[index + 1]);
+                index = index + 2;
+            }
+
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * @return a new list of all keys (makes a copy)
+     */
+    public Set<LongPair> items() {
+        Set<LongPair> items = new HashSet<>();
+        forEach((item1, item2) -> items.add(new LongPair(item1, item2)));
+        return items;
+    }
+
+    /**
+     * @return a new list of keys with max provided numberOfItems (makes a copy)
+     */
+    public Set<LongPair> items(int numberOfItems) {
```

Comment:
I think we will need `items(limit)` because `forEachLimit()` will hold a mutex until application processes all the items in that iteration. So, application/Dispatcher requires copying item using items(limit) and processes them as it wants. so, `items(limit)` has its usecase.

I have addressed all the changes except this one. if you still think it doesn't make sense then we can remove it and might introduce later.?